### PR TITLE
Implement local challenge updates (#57)

### PR DESCRIPTION
**Pull request by NicholasG04**
Implements #57 whilst allowing the flexibility to add any future local modifications. Simply takes in challenge data in the standard form (hopefully), and updates local state. Event Type 7 has been used.

Regards,
Nick

### CODE REVIEW

**Review by JoshHeng — APPROVED** [review]
LGTM!

File: src/@ractf/api/challenge.js [hunk 1]
```diff
@@ -36,6 +36,12 @@ export const resyncChallenge = (id) => {
     });
 };
 
+export const updateChallengeLocal = (data) => {
+    // data in the form { challenge_id: number, score: number, name: string } etc
+    store.dispatch(actions.editChallenge(data));
+    return data;
```
Comment (AmusedNetwork) [inline src/@ractf/api/challenge.js:42]: Why does this return `data` if the only usage doesn't use the return value? It is highly likely I don't understand JavaScript.

File: src/plugins/websocketLocalUpdate.js [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+
+import { registerPlugin } from "@ractf/plugins";
+import { updateChallengeLocal } from "@ractf/api";
+
+
+const WS_CHALLENGE_LOCAL_UPDATE = 7;
+
+export default () => {
+    registerPlugin("wsMessage", WS_CHALLENGE_LOCAL_UPDATE, (data) => {
+        updateChallengeLocal(data);
```
Comment (AmusedNetwork) [inline src/plugins/websocketLocalUpdate.js:26]: Is this just calling updateChallengeLocal, is the return value from updateChallengeLocal returned?

**Review by AmusedNetwork — CHANGES_REQUESTED** [review]
There's a minor issue on the copyright header, and I have some questions of my own that don't necessarily require changes.

File: src/@ractf/api/challenge.js (hunk 1 above)
Comment (NicholasG04) [reply src/@ractf/api/challenge.js:42]: good question, the rest of them do and this codebase doesn't use TypeScript so I'm too scared not to!

File: src/plugins/websocketLocalUpdate.js (hunk 2 above)
Comment (NicholasG04) [reply src/plugins/websocketLocalUpdate.js:26]: yes, but it's unused - we don't need it

File: src/plugins/websocketLocalUpdate.js [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with RACTF.  If not, see <https://www.gnu.org/licenses/>.
+
+import { registerPlugin } from "@ractf/plugins";
+import { updateChallengeLocal } from "@ractf/api";
+
+
+const WS_CHALLENGE_LOCAL_UPDATE = 7;
```
[the remark this answers is not in the document]
Comment (NicholasG04) [reply src/plugins/websocketLocalUpdate.js:22]: fair comment, will add

**Comment by david-cooke** [comment]
Changes on core:
 - Dynamic score challenges have a `current_score` attribute, when this is present & not null this should be displayed over `score`
 - `points_type` can now be modified by admins, this should be either `basic` or `decay`(probably should be a dropdown box for future extensibility
 - `flag_metadata.min_score` and `flag_metadata.decay_constant` can be configured per challenge, they will default to 100 and 0.99

**Comment by NicholasG04** [comment]
Many thanks. I'll look to implement these changes today.

**Review by thebeanogamer — CHANGES_REQUESTED** [review]
The correct score is being shown on the details page

image

But not the overview

image

Or the challenge admin overview

image

**Comment by thebeanogamer** [comment]
Websockets do seem to work though

image

**Review by thebeanogamer — APPROVED** [review]
image

**Pull request merged**